elytron-oidc-client/subsystem/src/main/java/org/wildfly/extension/elytron/oidc/OidcConfigService.java

```diff
@@ -0,0 +1,310 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2021 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.extension.elytron.oidc;
+
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.ADDRESS;
+import static org.wildfly.extension.elytron.oidc._private.ElytronOidcLogger.ROOT_LOGGER;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.jboss.as.controller.SimpleAttributeDefinition;
+import org.jboss.as.server.deployment.DeploymentUnit;
+import org.jboss.as.web.common.WarMetaData;
+import org.jboss.dmr.ModelNode;
+import org.jboss.dmr.Property;
+import org.jboss.metadata.web.jboss.JBossWebMetaData;
+
+public final class OidcConfigService {
+
+    private static final String CREDENTIALS_JSON_NAME = "credentials";
+
+    private static final String REDIRECT_REWRITE_RULE_JSON_NAME = "redirect-rewrite-rules";
+
+    private static final OidcConfigService INSTANCE = new OidcConfigService();
```

Comment:
This use of a static makes me nervous. I'd be a lot less so if this had a clear method and ElytronOidcSubsystemAdd called it.

This is kind of a code hardening thing. The DUPs presumably remove state on undeploy. A reload would not remove state added by an OSH, but OTOH for that state to be invalid left-behind cruft there would need to be a remove op that was called before the reload. But that's fragile and I bet I can dream up use cases where it breaks.

Reply:
jira issue for this and rest of the comments created: https://issues.redhat.com/browse/WFLY-15293